Core/src/Geometry/GridPortalLink.cpp

```diff
@@ -0,0 +1,421 @@
+    }
+
+    // If full cylinder, make sure axis wraps around
+    if (same(hlPhi, M_PI)) {
+      if (axis.getBoundaryType() != AxisBoundaryType::Closed) {
+        throw std::invalid_argument(
+            "GridPortalLink: CylinderBounds: invalid phi sector setup: "
+            "axis is not closed.");
+      }
+    } else {
+      if (axis.getBoundaryType() == AxisBoundaryType::Closed) {
+        throw std::invalid_argument(
+            "GridPortalLink: CylinderBounds: invalid phi sector setup: "
+            "axis is closed.");
+      }
+    }
+  };
+
+  if (dim() == 1) {
+    const IAxis& axisLoc0 = *grid().axes().front();
+    if (direction() == BinningValue::binRPhi) {
+      checkRPhi(axisLoc0);
+    } else {
+      checkZ(axisLoc0);
+    }
+  } else {  // DIM == 2
+    const auto& axisLoc0 = *grid().axes().front();
+    const auto& axisLoc1 = *grid().axes().back();
+    checkRPhi(axisLoc0);
+    checkZ(axisLoc1);
+  }
+}
+
+void GridPortalLink::checkConsistency(const DiscSurface& disc) const {
+  constexpr auto tolerance = s_onSurfaceTolerance;
+  auto same = [](auto a, auto b) { return std::abs(a - b) < tolerance; };
+
+  const auto* bounds = dynamic_cast<const RadialBounds*>(&disc.bounds());
+  if (bounds == nullptr) {
+    throw std::invalid_argument(
+        "GridPortalLink: DiscBounds: invalid bounds type.");
+  }
+
+  if (bounds->get(RadialBounds::eAveragePhi) != 0) {
+    throw std::invalid_argument(
+        "GridPortalLink: DiscBounds: only average phi == 0 is supported. "
+        "Rotate the disc surface.");
+  }
+
+  auto checkR = [&](const IAxis& axis) {
+    ActsScalar minR = bounds->get(RadialBounds::eMinR);
+    ActsScalar maxR = bounds->get(RadialBounds::eMaxR);
+    if (!same(axis.getMin(), minR) || !same(axis.getMax(), maxR)) {
+      throw std::invalid_argument(
+          "GridPortalLink: DiscBounds: invalid radius setup.");
+    }
+  };
+
+  auto checkPhi = [&](const IAxis& axis) {
+    ActsScalar hlPhi = bounds->get(RadialBounds::eHalfPhiSector);
+    if (!same(axis.getMin(), -hlPhi) || !same(axis.getMax(), hlPhi)) {
+      throw std::invalid_argument(
+          "GridPortalLink: DiscBounds: invalid phi sector setup.");
+    }
+    // If full disc, make sure axis wraps around
+    if (same(hlPhi, M_PI)) {
+      if (axis.getBoundaryType() != Acts::AxisBoundaryType::Closed) {
+        throw std::invalid_argument(
+            "GridPortalLink: DiscBounds: invalid phi sector setup: axis is "
+            "not closed.");
+      }
+    } else {
+      if (axis.getBoundaryType() == Acts::AxisBoundaryType::Closed) {
+        throw std::invalid_argument(
+            "GridPortalLink: DiscBounds: invalid phi sector setup: axis "
+            "is closed.");
+      }
+    }
+  };
+
+  if (dim() == 1) {
+    const IAxis& axisLoc0 = *grid().axes().front();
+    if (direction() == BinningValue::binR) {
+      checkR(axisLoc0);
+    } else {
+      checkPhi(axisLoc0);
+    }
+  } else {  // DIM == 2
```

Comment:
I guess we will never have more than 2 dimensions?

Reply:
That's the plan, yes.